sound/soc/sof/trace.c

```diff
@@ -167,7 +167,9 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 	struct sof_ipc_reply ipc_reply;
 	int ret;
 
-	if (sdev->dtrace_is_enabled || !sdev->dma_trace_pages)
+	if (!sdev->dtrace_is_enabled)
+		return 0;
```

Comment:
@plbossart do we need this check? we wont ever get here if trace is not enabled isnt it?

Reply:
@plbossart sorry I got confused here. This check seems incorrect. I think the previous one was correct. snd_sof_init_trace_ipc() should return -EINVAL is dtrace_is_enabled is already true

Reply:
I think we should use:
```
if (!IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_FIRMWARE_TRACE) &&
	    (sof_core_debug & SOF_CORE_ENABLE_TRACE) == 0) {
		return 0;
	}
```

Reply:
I think we need to check both for dma_trace_pages+dtrace_is_enabled and a check for whether trace-config option is enable.

First is needed because tracing might be enabled, but the allocation can fail and this is not treated as a fatal error. So we e.g. the pagetables might not have been created. We can not go enabling tracing towards FW if this is the case.

Second, you need to check for whether trace-config is enabled as this function is called from pm.c -> will be called even if boot flow did not create dtrace.

Reply:
I don't understand the assumptions made by the PM stuff. Is the field 'dtrace_is_enabled' ever set to false on suspend? are the DMA trace pages released on suspend?

Reply:
@plbossart yes we call 	snd_sof_release_trace() during suspend. But the DMA trace pages are not released.

Reply:
@plbossart The pages check is for "survivability". DMA allocs can fail but this is considered non-lethal now and other code needs to be able to handle this (e.g. in PM paths we need to ignore DMA trace if it was not properly initialized).

The "dtrace_is_enabled" check is more superfluous. Other functions called from PM do not check for this, but rather assume suspend is not called twice.

Reply:
Thanks. the module variable is static in core.c and exposing it as a global to test it in trace.c would be odd. The only solution I see is to add a 'trace_supported" boolean that's set when the Kconfig or module option are set.